### PR TITLE
Custom unicode handling for Android backspace via JNI to ICU

### PR DESCRIPTION
**Pull request by alimahdiyar**
The android backspace key is currently handled by TextKeyListener from native Android API in my older pull request #17393.
I figured out that this function has problems on lower Android API levels and I also think flutter needs a method to handle this functionality by itself.
So started scraping emoji characters from http://www.unicode.org/reports/tr51/#emoji_data by a python script and handled emojis based on the Unicode's documentation.
The utility I wrote is supporting all characters in Emoji 13.0

I also handled some states that were not handled in the android source code.

### CODE REVIEW

**Comment by GaryQian** [comment]
Can you file an issue to fully explain the problem? This would help with both understanding this and for future archeology to figure out what was going on.

We may want to consider using the Emoji handling that is already included in the ICU library that the C++ engine imports. This would ensure that we are using a standardized solution to this problem and that we don't have repeated logic in Java and C++. We should explore the tradeoffs of a direct custom java solution like yours vs a JNI into C++ to use ICU solution.

**Comment by alimahdiyar** [comment]
I started an emulator with Android 5.1 and wasn't able to delete characters with TextKeyListener.

icu for java requires Android API level 21.
http://site.icu-project.org/download/66#TOC-ICU4J-Download

I saw the android source code is still handling some new Emojis alongside icu library.
https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/text/Emoji.java#75

How should I compare performance between the two solutions? (using JNI - my Java code)
@GaryQian

**Comment by justinmc** [comment]
I am able to delete complex emojis like 👨‍👩‍👦 using the existing implementation with `TextKeyListener.getInstance().onKeyDown`.  I'm using an Android Pixel 3 emulator with API 28. 
 Can you post specific steps to reproduce?  Maybe open an issue in the framework repository like Gary said, or find an existing issue that covers it.

I'll defer to Gary on the C++ and performance questions.

Thanks for opening this by the way!  It may unblock my PR for Dpad navigation through emoji characters (https://github.com/flutter/engine/pull/17420) if we can figure that out.

**Comment by jason-simmons** [comment]
I can reproduce the issue using an emulator running an API 21 (Lollipop) image.  In that environment the `TextKeyListener.onKeyDown` based implementation does not remove any text when the delete key is pressed.

I think the `getOffsetBefore` approach is a good improvement, but I'd like to avoid reimplementing a grapheme cluster breaker in the Android embedding.

Could the `InputConnectionAdaptor` call Android's `android.text.TextUtils.getOffsetBefore` on older platforms and use the `android.icu.text.BreakIterator` on platforms that offer it (API 24 and later)?
`TextUtils.getOffsetBefore` will not handle the latest emoji, but it should provide accuracy equivalent to native text fields on the Android versions where it would be used.

The JNI-based alternative would involve adding a new JNI exported function to the engine that provides an API like `getOffsetBefore`.  The native implementation would use the grapheme breaker in the ICU C++ library that is already linked into the Flutter engine.  I'd prefer to reuse that if there is no sufficiently accurate breaker that is already available to Java.

**Comment by alimahdiyar** [comment]
I don't trust Android APIs for this specific problem as we encountered problems and if we are about to use something to avoid implementing, I prefer using JNI solution.
Can you please give me some hint on which functions should I use from the native engine and how to export them?

**Comment by alimahdiyar** [comment]
So I'm using JNI and icu module from the native engine.
I tested on emulators and code worked fine but in flutter tests I got:
```
java.lang.UnsatisfiedLinkError: io.flutter.plugin.editing.EmojiUtils.nativeEmojiUtilsIsRegionalIndicator(I)Z
	at io.flutter.plugin.editing.EmojiUtils.nativeEmojiUtilsIsRegionalIndicator(Native Method)
```
And I don't know why. Can you help me with this, please?

**Comment by GaryQian** [comment]
You should look at `FlutterJNI.java`. We currently implement all JNI methods in there, an explanation is given in the docs at the top.

I am not familiar enough with JNI to be able to tell you what is immediately wrong just from looking, but in any case, to stay consistent with the rest of the engine, you should move this into FlutterJNI.java anyways. There should be plenty of example methods already implemented in there to work off of.

**Comment by alimahdiyar** [comment]
I tried flutter JNI instance. still the same error

**Comment by GaryQian** [comment]
Hmm, which tests are you running exactly to encounter that?

**Comment by alimahdiyar** [comment]
InputConnectionAdaptor last two tests.
It is also failing with the same error in Cirrus CI Linux Android Debug Engine.

**Comment by jason-simmons** [comment]
I posted a version of this that finds the character boundary by using the `icu::BreakIterator` through JNI:
  https://github.com/flutter/engine/pull/18041

Can you try it and see if it works for your use case?

**Comment by alimahdiyar** [comment]
Tests are now working with a solution from #18041.
I think now PR is in a good state.

**Comment by alimahdiyar** [comment]
Tests are now working with a solution from #18041.
I think now PR is in a good state.
Please review again.
@GaryQian @jason-simmons

**Comment by alimahdiyar** [comment]
I figured out `mock` should be used for these tests and the linking problem is solved now.
The code is fine on real devices.
But the native functions are all returning false in tests and I don't know why.
Can you help me figure out why?

**Comment by jason-simmons** [comment]
The mock implementation of `FlutterJNI` will return false by default in all of its methods.

Use Mockito APIs to provide replacement handlers for calls to the nativeTextUtils methods:

```
    when(mockFlutterJni.nativeTextUtilsIsEmoji(anyInt())).thenAnswer(
      (invocation) -> Emoji.isEmoji((int) invocation.getArguments()[0])
    );
    when(mockFlutterJni.nativeTextUtilsIsEmojiModifier(anyInt())).thenAnswer(
      (invocation) -> Emoji.isEmojiModifier((int) invocation.getArguments()[0])
    );
    when(mockFlutterJni.nativeTextUtilsIsEmojiModifierBase(anyInt())).thenAnswer(
      (invocation) -> Emoji.isEmojiModifierBase((int) invocation.getArguments()[0])
    );
    when(mockFlutterJni.nativeTextUtilsIsVariationSelector(anyInt())).thenAnswer(
      (invocation) -> {
        int codePoint = (int) invocation.getArguments()[0];
        return 0xFE0E <= codePoint && codePoint <= 0xFE0F;
      }
    );
    when(mockFlutterJni.nativeTextUtilsIsRegionalIndicator(anyInt())).thenAnswer(
      (invocation) -> Emoji.isRegionalIndicatorSymbol((int) invocation.getArguments()[0])
    );
```

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 1]
```diff
@@ -263,26 +341,153 @@ private static int clampIndexToEditable(int index, Editable editable) {
     return clamped;
   }
 
+  public int getOffsetBefore(int offset) {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:344]: Can you move this into a new `FlutterTextUtils.java` file?

I'd like to keep general purpose text utilities separate from the rest of `InputConnectionAdaptor` which handles interaction with Android's text input services.

Also add a comment explaining what `getOffsetBefore` does and why Android's similar APIs are insufficient.

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 2]
```diff
@@ -263,26 +341,153 @@ private static int clampIndexToEditable(int index, Editable editable) {
...
+    if (offset <= 1) {
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:345]: I would also provide a link to Android's implementation as reference for an existing version of this logic. https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android10-s3-release/core/java/android/text/method/BaseKeyListener.java#111 Seems like a good version to link to. Master may change so this is the android 10 version.

Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:345]: This is also in Chromium's Blink: https://source.chromium.org/chromium/chromium/src/+/84.0.4131.5:third_party/blink/renderer/core/editing/state_machines/backspace_state_machine.cc

Comment (alimahdiyar) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:345]: My refrence for handling this behavior was the android source code, too.

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java (hunk 1 above)
Comment (alimahdiyar) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Done

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+package io.flutter.plugin.editing;
+
+import io.flutter.embedding.engine.FlutterJNI;
+
+class FlutterTextUtils {
+  public static final int LINE_FEED = 0x0A;
+  public static final int CARRIAGE_RETURN = 0x0D;
+  public static final int COMBINING_ENCLOSING_KEYCAP = 0x20E3;
+  public static final int CANCEL_TAG = 0xE007F;
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:13]: I believe the Unicode spec refers to this as `TAG_END`, should probably stick to same naming scheme

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+          deleteCharCount += Character.charCount(codePoint) + Character.charCount(tmpCodePoint);
+        }
+      } else if (isKeycapBase(codePoint)) {
+        deleteCharCount += Character.charCount(codePoint);
+      }
+      return offset - deleteCharCount;
+    }
+
+    // Emoji Tag Sequence
+    if (codePoint == CANCEL_TAG) { // tag_base
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:111]: did you mean `// tag_end`?

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+      codePoint = Character.codePointBefore(text, lastOffset);
+      lastOffset -= Character.charCount(codePoint);
+      while (lastOffset > 0 && isTagSpecChar(codePoint)) { // tag_spec
+        deleteCharCount += Character.charCount(codePoint);
+        codePoint = Character.codePointBefore(text, lastOffset);
+        lastOffset -= Character.charCount(codePoint);
+      }
+      if (lastOffset == 0) { // tag_end not found. Just delete the end.
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:119]: did you mean `tag_base`?

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java (hunk 3 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:13]: http://unicode.org/reports/tr51/#def_emoji_tag_sequence

Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:13]: Jk, it looks like they do call this `CANCEL_TAG`, this is fine.

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        int tmpCodePoint = Character.codePointBefore(text, lastOffset);
+        if (isKeycapBase(tmpCodePoint)) {
+          deleteCharCount += Character.charCount(codePoint) + Character.charCount(tmpCodePoint);
+        }
+      } else if (isKeycapBase(codePoint)) {
+        deleteCharCount += Character.charCount(codePoint);
+      }
+      return offset - deleteCharCount;
+    }
+
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:109]: Since you have unrolled the state machine, can you add some notation here indicating that the following few if statements are meant to act like a fall through switch and do not return like the if statements above?

The separation will make the code much more parse-able.

File: shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java [hunk 7]
```diff
@@ -334,9 +337,85 @@ public void testSendKeyEvent_delKeyDeletesBackward() {
...
+
+    didConsume = adaptor.sendKeyEvent(downKeyDown);
+    assertTrue(didConsume);
+    assertEquals(Selection.getSelectionStart(editable), 0);
+  }
+
   private static final String SAMPLE_TEXT =
       "Lorem ipsum dolor sit amet," + "\nconsectetur adipiscing elit.";
 
+  private static final String SAMPLE_EMOJI_TEXT =
```
Comment (GaryQian) [inline shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java:405]: You should add some additional edge cases, including handling of invalid sequences. Invalid sequences may be generated programmatically, and we should make sure we can handle/recover from these cases.

**Comment by GaryQian** [comment]
Implementation roughly LGTM, will take another look during final pass.

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java (hunk 3 above)
Comment (alimahdiyar) [reply shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:13]: I think it's better to follow the Android guideline because we are using it as the reference.

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java (hunk 5 above)
Comment (alimahdiyar) [reply shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:119]: Was a mistake. fixed.

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java (hunk 4 above)
Comment (alimahdiyar) [reply shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:111]: Was a mistake. fixed.

File: shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java (hunk 6 above)
Comment (alimahdiyar) [reply shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java:109]: Done

File: shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java (hunk 7 above)
Comment (alimahdiyar) [reply shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java:485]: Done

**Comment by alimahdiyar** [comment]
I handled invalid sequences and unknown cases in the last changes
